movement/watch_faces/complication/totp_face_lfs.c

```diff
@@ -0,0 +1,231 @@
+            return false;
+        }
+    } else if (!strcmp(param, "digits")) {
+        if (!strcmp(param, "6")) {
+            printf("TOTP got %s, not 6 digits\n", value);
+            return false;
+        }
+    } else if (!strcmp(param, "period")) {
+        totp_record->period = atoi(value);
+        if (totp_record->period == 0) {
+            printf("TOTP invalid period %s\n", value);
+            return false;
+        }
+    } else if (!strcmp(param, "algorithm")) {
+        if (!strcmp(param, "SHA1")) {
+            printf("TOTP ignored due to algorithm %s\n", value);
+            return false;
+        }
+    }
+
+    return true;
+}
+
+static void totp_face_lfs_read_file(char *filename) {
+    // For 'format' of file, see comment at top.
+    const size_t uri_start_len = strlen(TOTP_URI_START);
+
+    if (!filesystem_file_exists(filename)) {
+        printf("TOTP file error: %s\n", filename);
+        return;
+    }
+
+    char line[256];
+    int32_t offset = 0;
+    while (filesystem_read_line(filename, line, offset, 255) && strlen(line)) {
+        offset += strlen(line) + 1;
+        
+        if (num_totp_records == MAX_TOTP_RECORDS) {
+            printf("TOTP max records: %d\n", MAX_TOTP_RECORDS);
+            break;
+        }
+
+        // Check that it looks like a URI
+        if (strncmp(TOTP_URI_START, line, uri_start_len)) {
+            printf("TOTP invalid uri start: %s\n", line);
+            continue;
+        }
+
+        // Check that we can find a '?' (to start our parameters)
+        char *param;
+        char *param_saveptr = NULL;
+        char *params = strchr(line + uri_start_len, '?');
+        if (params == NULL) {
+            printf("TOTP no params: %s\n", line);
+            continue;
+        }
+
+        // Process the parameters and put them in the record
+        init_totp_record(&totp_records[num_totp_records]);
+        bool error = false;
+        param = strtok_r(params + 1, "&", &param_saveptr);
+        do {
+            char *param_middle = strchr(param, '=');
+            *param_middle = '\0';
+            error = error || !totp_face_lfs_read_param(&totp_records[num_totp_records], param, param_middle + 1);
+        } while ((param = strtok_r(NULL, "&", &param_saveptr)));
+
+        if (error) {
+            totp_records[num_totp_records].secret_size = 0;
+            continue;
+        }
+
+        // If we found a probably valid TOTP record, keep it. 
+        if (totp_records[num_totp_records].secret_size) {
+            num_totp_records += 1;
+        } else {
+            printf("TOTP missing secret: %s\n", line);
+        }
+    }
+}
+
+void totp_face_lfs_setup(movement_settings_t *settings, uint8_t watch_face_index, void ** context_ptr) {
+    (void) settings;
+    (void) watch_face_index;
+    if (*context_ptr == NULL) {
+        *context_ptr = malloc(sizeof(totp_lfs_state_t));
+        totp_face_lfs_read_file(TOTP_FILE);
```

Comment:
Theoretically, to support multiple TOTP faces, I guess this should be in a separate 'if' to check if our static data has been initialised. But in practice it would only make sense to have multiple faces if we had different TOTP files, I think.

Reply:
yea I don't foresee multiple TOTP faces really making sense; since the file is hard coded, I think it's reasonable to say "have a maximum of one TOTP face in the list"